packages/react-dom/src/events/ReactDOMEventListener.js

```diff
@@ -189,12 +193,15 @@ export function dispatchEvent(
   topLevelType: DOMTopLevelEventType,
   nativeEvent: AnyNativeEvent,
 ) {
-  if (!_enabled) {
+  if (!_enabled || hasEventDispatched(nativeEvent)) {
     return;
   }
 
+  trackEventDispatch(nativeEvent);
```

Comment:
This is the new logic: do not dispatch the same native event to React more than once. This prevents the following case:


```
<div onScroll={...}>
  <div onScroll={...}>
</div>
```

If a scroll event on the child triggers:

1. The child's internal React event handler dispatches to the synthetic event system
2. The child callback fires
3. The parent callback fires
4. The native event bubbles to the parent
5. The parent's internal React event handler dispatches to the synthetic event system
6. The parent callback fires

The expected behavior is that the child and parent callback are only triggered once, however the parent will trigger _twice_.

By tracking if an event has already dispatched, we can bail early and preserve the expected behavior

Reply:
What if you intentionally `.dispatch()` the same native event twice? e.g. in testing?

Reply:
It wouldn't dispatch the second time. I wonder if you could get around that by removing the event from the set when it hits the root element (or some other indicator that the event was "done" dispatching).

Reply:
Calling `event.stopPropagation()` might mean that it never hit the root. What if we store a list of all elements that the event already targeted and reset once a duplication is found?

I think the problem with that is that we would ignore the second dispatch if the same event targets another subtree 🤔

```
<div id="a" onScroll={...} />
<div id="b" onScroll={...} />

a.dispatchEvent(event);
b.dispatchEvent(event);
```

Now that I understand what's the issue here I can also wrap my head around that a bit 🙂

Reply:
Also let's keep in mind that the end game is that we attach event listeners at the root rather than document. If that would also solve this problem maybe it's a better fix to focus on?

Reply:
Something we could try for sure. I'm a little worried that we overestimate the power of per-root event listener though. We seem to praise it as the solution for all of our event problems right now.

At some point I want to compile a list of issues that we've found to see which of those per-root listening can solve and what other problems might occur.

Reply:
Well. I think in this case the original issue was https://github.com/facebook/react/issues/1254 which is called "React shouldn't bind wheel or touch events to the document". So that does sound to me as a specific instance of this more generic problem. :-)

Reply:
> Also let's keep in mind that the end game is that we attach event listeners at the root rather than document. If that would also solve this problem maybe it's a better fix to focus on?

I gave that a shot, but I wasn't able to prevent scroll jank. Attaching listeners locally was the only way that I could prevent it. I can do some more testing just to be sure, though.

Reply:
Also, if this is _only_ for scroll events, I wonder if we need to synthetically bubble/capture at all. For events that don't have synthetic dispatch phases (like mouse and composition events), maybe DOM bubbling/capture is simply sufficient.

Reply:
>I gave that a shot, but I wasn't able to prevent scroll jank. Attaching listeners locally was the only way that I could prevent it. I can do some more testing just to be sure, though.

For which events? Did you include wheel, scroll, and touch events?

Reply:
> Doesn't implement portal bubbling.

Ah shoot 😢. 

> For which events? Did you include wheel, scroll, and touch events?

Wheel and scroll. But I'll give it another shot.

Reply:
Touch events seem essential to this.

Reply:
Portal bubbling is curious to me. Portals can exist in a different tree than the parent DOM elements, but still accept bubbled events, like clicks?

Reply:
https://reactjs.org/docs/portals.html#event-bubbling-through-portals

Reply:
Scroll is also a requirement for the responder event plugin so that might break  RNW if we remove it from the synthetic event system.

Reply:
I sent out an alternative PR that attaches to the root here:
https://github.com/facebook/react/pull/13460

This might work, I wrote the idea off because of some issues with mouse input on BrowserStack (I think, anyway). This definitely needs testing, but the approach in #13460 definitely feels ideal if I can figure out issues in Safari.